tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -14161,6 +14161,97 @@ static bool IsUsageAttributeCompatible(AttributeList::Kind kind, bool &usageIn,
   return true;
 }
 
+bool DiagnoseRayQueryObject(QualType qt, Declarator &D, Sema &S) {
+  // now we know that qt is a TemplateSpecializationType
+  // handle diagnostics that relate to enumtypedecl vs literal first
+  const TemplateSpecializationType *TST =
+      dyn_cast<TemplateSpecializationType>(qt);
+  if (!TST) {
+    return false;
+  }
+
+  auto typeRecordDecl = qt->getAsCXXRecordDecl();
+  ClassTemplateSpecializationDecl *SpecDecl =
+      llvm::dyn_cast<ClassTemplateSpecializationDecl>(typeRecordDecl);
+  if (!SpecDecl) {
+    return false;
+  }
+
+  // get number of template args
+  unsigned numArgs = TST->getNumArgs();
+  assert((numArgs == 1 || numArgs == 2) &&
+         "All ray query declarations should have one or two template "
+         "args");
+  llvm::APSInt Arg1val = SpecDecl->getTemplateArgs()[0].getAsIntegral();
+  bool IsForceOMM2State =
+      Arg1val.getZExtValue() & (unsigned)DXIL::RayFlag::ForceOMM2State;
+
+  // if there's only one template argument, then it cannot be the
+  // ForceOMM2State flag, since the second argument needs to be
+  // DXIL::RayQueryFlag::AllowOpacityMicromaps, not the default 0
+  if (numArgs == 1) {
+    if (IsForceOMM2State) {
+      S.Diag(D.getLocStart(), diag::warn_hlsl_unexpected_rayquery_flag);
+      return false;
+    }
+    return true;
+  }
+
+  // we're guaranteed 2 args now
+
+  // ensure that if the second template argument has a non-zero value,
+  // the shader model is at least 6.9
+  const TemplateArgument *Args = TST->getArgs();
+  auto Arg2 = Args[1];
```

Comment:
Do you intend to create a copy of the template argument?

Maybe this should be:

```suggestion
  const TemplateArgument &Arg2 = Args[1];
```